Dpl degree switchover

davidornelas11 opened the pull request:
### Description
Updated app-degree-pages to use new data potluck api and corresponding logic

### Links

- [Unity reference site](https://unity.web.asu.edu/@asu/app-degree-pages/?path=/story/program-listing-page--default&acadPlan=PPSWDMSW)
- [JIRA ticket](https://asudev.jira.com/browse/UDS-1336?atlOrigin=eyJpIjoiMzYzMDM3NTE0NGVlNGI1N2IyYTFmZDM0Zjg2ZWI3NWMiLCJwIjoiaiJ9)

### Checklist

- [x] Unity project successfully builds from root `yarn install` & `yarn build`
- [x] Commits do not contain multiple scopes
- [x] Add/updated examples
- [x] Add/updated READMEs/docs
- [x] No new console errors
- [x] Accessibility checks

### Browsers

- [x] Chrome
- [x] Safari
- [x] Firefox
- [x] Edge

### Images

<!-- Provide screenshots -->

davidornelas11 committed:
feat(app-degree-pages): updated Detail page component

davidornelas11 committed:
feat(app-degree-pages): updated Listing Page Component

davidornelas11 committed:
feat(app-degree-pages): updated default api endpoint data

davidornelas11 committed:
feat(app-degree-pages): updating types and definitions for jsdocs

davidornelas11 committed:
feat(app-degree-pages): add new utils and updated for new parameters

davidornelas11 committed:
feat(app-degree-pages): updated resolver and filtering logic for degrees

davidornelas11 committed:
fix(app-degree-pages): update TODO's

scott-williams-az commented on packages/app-degree-pages/src/components/DetailPage/index.js:
```diff
@@ -255,12 +269,12 @@ const DetailPage = ({
+                    }
+                    concurrentLinks={
+                      formatAcceleratedConcurrentLinks(acceleratedAndConcurrentDegrees.concurrentData)
+                    }
```
Formatting, Lines too long

scott-williams-az commented on packages/app-degree-pages/src/components/ListingPage/components/ProgramList/ListView/index.colums.config.js:
```diff
@@ -117,10 +117,10 @@ const columns = [
     contentTemplate: ({ resolver, actionUrls }) => (
       <div>
-        {resolver.getConcurrentDegrees().lenght > 0 && (
+        {resolver.getConcurrentDegrees().then(data => data).length > 0 && (
```
Just double checking this function eventually calls fetchAcademicPlans which looks like it will alway return an array, do we need to handle if response is an error?

davidornelas11 replied:
I added error handling in the request itself so it will return an empty array, which fixes this issue

scott-williams-az commented on packages/app-degree-pages/src/core/utils/execute-promises-update-state.js:
```diff
@@ -0,0 +1,18 @@
+ */
+function executePromisesAndUpdateState(promises, stateUpdateFunction) {
+  Promise.all(promises)
+  .then(results => {
```
Fix formatting

scott-williams-az commented on packages/app-degree-pages/src/core/utils/fetch-plan-codes.js:
```diff
@@ -0,0 +1,23 @@
+ * @param {string[]} acadPlanCodes Array of academic plan codes.
+ * @returns {Promise<Object[]>} A promise that resolves to an array of academic plan details.
+ */
+async function fetchAcademicPlans(acadPlanCodes) {
```
Mentioned before, how are we handling errors? timeouts?

davidornelas11 replied:
Handling timeouts are always tricky, since I'm never sure how long to wait. I think for this instance, a 5s timeout for each request is decent. This is something we can talk about in standup. I will add a timeout request of 5s and we can  change it if we need to

mlsamuelson replied:
5s, but maybe 10s would lead to less likelihood of false positives while still having a safeguard.

scott-williams-az commented on packages/app-degree-pages/src/core/utils/http-url-resolver.js:
```diff
@@ -17,6 +17,25 @@ function urlResolver(dataSource, defaultDataSource) {
+    delete httpParameters["filter"];
+  }
 
   const { endpoint, fields, ...keyValues } = httpParameters;
```
fields is no longer being used

scott-williams-az commented on packages/app-degree-pages/src/core/utils/string-utils.js:
```diff
@@ -1,6 +1,7 @@
 // @ts-check
 
 function toTitleCase(str) {
+  if (typeof str !== "string") return console.error("Invalid string", str);
```
return statement like this I try to avoid since they are easier to overlook. I thought we had a linting rule that prefers we do not do this shorthand. 

That being said since it is at the top of the function it doesn't bother me too much.

davidornelas11 committed:
fix(app-degree-pages): added wue check for campuses offered

mlsamuelson commented on packages/app-degree-pages/src/components/ListingPage/components/ProgramList/ListView/index.colums.config.js:
```diff
@@ -117,10 +117,10 @@ const columns = [
           <div className="cell-container">concurrent</div>
         )}
-        {resolver.getAccelerateDegrees().length > 0 && (
+        {resolver.hasConcurrentOrAccelerateDegrees() > 0 && (
```
Since `hasConcurrentOrAccelerateDegrees()` does a `!!` [on the length value and returns that](https://github.com/ASU/asu-unity-stack/blob/957cc815b0cbd70d6d90733dc9b2287720ad8434/packages/app-degree-pages/src/core/services/degree-data-prop-resolver-service.js#L85-L86), it would be better here to do a boolean rather than numerical comparison. It also makes more semantic sense, and improves readability since this is a "has" method.

davidornelas11 replied:
Good catch

mlsamuelson commented on packages/app-degree-pages/src/core/constants/web-api-constants.js:
```diff
@@ -6,55 +6,19 @@ const DOMAIN_URL = "https://degrees.apps.asu.edu";
-    "asuMathReqFlag,additionalMathReqCourse,asuAcadpMrfText,MathIntensity,",
+  endpoint: "https://api.myasuplat-dpl.asu.edu/api/codeset/acad-plans",
+  filter: "activeInDegreeSearch",
+  include: "*",
```
This is the default... And I haven't finished reviewing, but if we're not at some point confining the fields to include to only those we're using, that's likely to have a deleterious effect on performance.

Now that the conversion effort is done, can we replace the `*` include here and on line 21 below, with a request that only asks for fields we rely on?

davidornelas11 committed:
fix(app-degree-pages): address pr issues

davidornelas11 committed:
fix(app-degree-pages): fixed errors when building

mlsamuelson commented on packages/app-degree-pages/src/core/services/degree-data-prop-resolver-service.js:
```diff
@@ -32,36 +29,54 @@ const isValidActiveProgram = row =>
+    },
+    isUndergradProgram: () => isUndergradProgram(row),
+    isGradProgram: () => row["degreeType"] === "GR",
+    isMinorOrCertificate: () => row["degreeType"] === "UGCM",
```
We may need to account for Grad certificates. UGCM's UG is undergrad...

mlsamuelson commented on packages/app-degree-pages/src/core/services/degree-data-prop-resolver-service.js:
```diff
@@ -32,36 +29,54 @@ const isValidActiveProgram = row =>
+    getDegreeDesc: () => row["degreeDescriptionLong"],
+    getDegreeDescLong: () => row["degreeDescriptionText"],
+    getFullDescription: () => row["fullDescription"],
+    getCurriculumUrl: () => row["asuOnlineAcadPlanUrl"]?.trim(),
```
Just a note, that the curriculum URL was one of our fields to confirm with testing...

davidornelas11 committed:
fix(app-degree-pages): added logic for minorOrCert check

mlsamuelson commented on packages/app-degree-pages/src/core/utils/http-url-resolver.js:
```diff
@@ -17,16 +17,40 @@ function urlResolver(dataSource, defaultDataSource) {
+  }
+  if (httpParameters.cert === "true") {
+    httpParameters.degreeType = "UGCM";
+    delete httpParameters["cert"];
```
Dropping a comment here to find my way back, in case any adjustments may be needed based on Undergrad and Graduate cert handling differences.

mlsamuelson reviewed (commented):
I've done a first run through of the code in the PR. Will now be doing side-by-side testing.

davidornelas11 committed:
fix(app-degree-pages): addressed PR issues

davidornelas11 committed:
Merge branch 'dev' into dpl-degree-switchover

davidornelas11 committed:
fix(app-degree-pages): timeout for fetchAcadPlans to 10s

davidornelas11 committed:
feat(app-degree-pages): added stemOpt section to program detail page

davidornelas11 committed:
fix(app-degree-pages): fixed missing grad requirement info

mlsamuelson commented:
Reviewed results of side-by-side testing reviewed with @davidornelas11 and documented questions and next steps on https://asudev.jira.com/browse/UDS-1336?focusedCommentId=1933908 . Also, as noted on that ticket, we've reached out to Andrew Nelson with a question about how programs that aren't accepting new students...

davidornelas11 committed:
fix(app-degree-pages): subplan links now match old service

mlsamuelson commented on packages/app-degree-pages/src/components/ListingPage/components/ProgramList/ListView/index.colums.config.js:
```diff
@@ -67,11 +67,18 @@ const columns = [
-      );
+      let directMapLink;
+
+      try {
+        const url = new URL(directUrl);
+        directMapLink = (
+          <a href={directUrl}>{resolver.getRequiredCoursesLabel()} Map</a>
+        );
+      } catch (error) {
+        return "";
+      }
```
We need something like this logic implemented at https://github.com/ASU/asu-unity-stack/blob/dpl-degree-switchover/packages/app-degree-pages/src/components/ListingPage/components/ProgramList/AccordionView/index.js#L58 because the mobile and tablet accordion view includes the Required Courses major map link as an empty link for graduate. You'll probably want to include the label in that logic so it doesn't display when the link is absent.

davidornelas11 committed:
fix(app-degree-pages): added proptypes and removed section from graduate listing

Graduate listings should not have required courses since most will not ahve major maps
